File: codes/java/chap09/answer_A70.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		// 입력
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int M = sc.nextInt();
		int[] A = new int[N + 1];
		int[] X = new int[M + 1];
		int[] Y = new int[M + 1];
		int[] Z = new int[M + 1];
		for (int i = 1; i <= N; i++) {
			A[i] = sc.nextInt();
		}
		for (int i = 1; i <= M; i++) {
			X[i] = sc.nextInt();
			Y[i] = sc.nextInt();
			Z[i] = sc.nextInt();
		}

		// 인접 리스트를 만들고, 그래프에 에지를 추가
		ArrayList<Integer>[] G = new ArrayList[1 << N];
		for (int i = 0; i < (1 << N); i++) {
			G[i] = new ArrayList<Integer>();
		}
		for (int i = 0; i < (1 << N); i++) {
			for (int j = 1; j <= M; j++) {
				int nextState = getNext(N, i, X[j], Y[j], Z[j]);
				G[i].add(nextState);
			}
		}

		// 시작/골 노드 번호를 구한다
		int start = 0;
		for (int i = 1; i <= N; i++) {
			if (A[i] == 1) {
				start += (1 << (i - 1));
			}
		}
		int goal = (1 << N) - 1;

		// 배열 초기화／시작 시점을 큐에 넣는다
		int[] dist = new int[1 << N];
		Arrays.fill(dist, -1);
		dist[start] = 0;
		Queue<Integer> Q = new LinkedList<>();
		Q.add(start);

		// 너비 우선 탐색
		while (Q.size() >= 1) {
			int pos = Q.remove();
			for (int i = 0; i < G[pos].size(); i++) {
				int nex = G[pos].get(i);
				if (dist[nex] == -1) {
					dist[nex] = dist[pos] + 1;
					Q.add(nex);
				}
			}
		}

		// 답을 출력
		System.out.println(dist[goal]);
	}
	
	// 노드 pos의 상태에서 '램프 x, y, z의 상태'를 반전시켰을 때의 노드 번호를 반환하는 함수
	// (책의 코드에서는 '노드 pos의 상태에서 idx번째 종류의 조작을 수행했을 때'의 노드 번호를 반환하는 함수를 구현해고 있지만, 이것과는 다소 다른 구현을 한 것에 주의)
	static int getNext(int N, int pos, int x, int y, int z) {
		// pos의 2진법 표기를 사용해, 노드 pos가 나타내는 램프의 상태 state를 계산
		// (2진법으로 변환하는 방법은 1.4절을 참조)
		int[] state = new int[N + 1];
		for (int i = 1; i <= N; i++) {
			int wari = (1 << (i - 1));
			state[i] = (pos / wari) % 2;
		}

		// 램프 x, y, z의 상태를 반전
		state[x] = 1 - state[x];
		state[y] = 1 - state[y];
		state[z] = 1 - state[z];

		// 10진법으로 변환하는 방법도 1.4절을 참조
		int ret = 0;
		for (int i = 1; i <= N; i++) {
			if (state[i] == 1) {
				ret += (1 << (i - 1));
			}
		}
		return ret;
	}
}

// 주의 1: 이 문제에 대해서는 보다 간략한 구현도 있으므로, 필요하다면 answer_A70_extra.py를 참조합니다.

File: codes/python/chap09/answer_A70.py
from collections import deque

# 입력(여기에서는 책과 달리 램프의 번호가 0-indexed가 되도록 구현하고 있다)
N, M = map(int, input().split())
A = list(map(int, input().split()))
actions = [ list(map(lambda x: int(x) - 1, input().split())) for i in range(M) ] # ここでは X[i], Y[i], Z[i] を 0-indexed に変換して受け取る

# 노드 pos의 상태에서 '램프 x, y, z의 상태'를 반전시켰을 때의 노드 번호를 반환하는 함수
def get_next(pos, x, y, z):
	# pos의 2진법 표기를 사용해, 노드 pos가 나타내는 램프의 상태 state를 계산
	# （pos の 2^i の位は (pos // (2 ** i)) % 2 で計算できる → 1.4 節を参照）
	state = [ (pos // (2 ** i)) % 2 for i in range(N) ]
	# ランプ x, y, z を反転
	state[x] = 1 - state[x]
	state[y] = 1 - state[y]
	state[z] = 1 - state[z]
	# ランプの状態 state を指す頂点の番号を計算
	# （2 進法を 10 進法に変換する方法は 1.4 節を参照）
	ret = 0
	for i in range(N):
		if state[i] == 1:
			ret += 2 ** i
	return ret

# グラフに辺を追加
G = [ list() for i in range(2 ** N) ]
for i in range(2 ** N):
	for x, y, z in actions:
		nextstate = get_next(i, x, y, z)
		G[i].append(nextstate)

# 시작 지점/골 지점의 노드 번호를 결정한다
start = 0
for i in range(N):
	if A[i] == 1:
		start += 2 ** i
goal = 2 ** N - 1

# 너비 우선 탐색 초기화
dist = [ -1 ] * (2 ** N)
dist[start] = 0
Q = deque()
Q.append(start)

# 너비 우선 탐색
while len(Q) >= 1:
	pos = Q.popleft() # 큐 Q의 맨 앞 요소를 제거하고, 그 값을 pos에 대입한다
	for nex in G[pos]:
		if dist[nex] == -1:
			dist[nex] = dist[pos] + 1
			Q.append(nex)

# 답을 출력
print(dist[goal])

# 注意 1：この問題に対してはより簡潔な実装もありますので、
#         もしよければ answer_A70_extra.py もご覧ください。